### DISCUSSION TITLE
Filtering DB side with TimeOnly properties.

### DISCUSSION THREAD

**Discussion by dtaalbers** [ravendb/ravendb#17560] [Q&A]
Hi there,

I have another issue that I don't really understand. I have the question mocked in [this example](https://dotnetfiddle.net/LN4XJA).

Basically I have `Schedules` with a list of `SchedulePeriod`. No

```
public class Schedule
{
	public string Id { get; set; }
	public string UserId { get; set; }
	public List<SchedulePeriod> Periods { get; set; }
}

public class SchedulePeriod
{
	public DayOfWeek Day { get; set; }
	public TimeOnly Start { get; set; }
	public TimeOnly End { get; set; }
}
```

When creating an appointment I want to check if that appointment falls within any schedule of a user. Now with the example I provided I am almost there. The weird thing is that when the end time of the appointment is `12:00` it doesn't find the schedule when filtering on the database side. It does however find the schedule when I fetch all schedules first, and then filter with the same filter requirements. Is this a bug or am I doing something wrong? Any help would be appreciated.

I have also read [this](https://ravendb.net/docs/article-page/6.0/csharp/client-api/how-to/using-timeonly-and-dateonly#using-already-existing-dateonly-or-timeonly-fields). Which I tried to apply in the example. But I am not sure if I have done that right.

```
public class Appointment
{
	public string Id { get; set; }
	public DateTime Start { get; set; }
	public DateTime End { get; set; }
	public string User { get; set; }
}
```

**Comment by ayende** [comment]
Note that your index is not used (and won't actually do what you want it to do). 
We don't inspect the internal of objects as we index them.

As for the actual problem, it looks like when we index the value, we use `12:00:00` and when we query, we use `12:00:00.0000000`, which is why you are running into this issue:

https://issues.hibernatingrhinos.com/issue/RavenDB-21588/Range-and-equality-check-on-TimeOnly-values

**Reply by garayx** [reply]
it seems that the issue is only on lucene,
@dtaalbers  can you switch the search engine to corax and see if it return the expected results ?
https://ravendb.net/docs/article-page/6.0/csharp/indexes/search-engine/corax#selecting-the-search-engine

**Reply by dtaalbers** [reply]
@garayx @ayende 

Thanks for the reply! I've updated [the example](https://dotnetfiddle.net/LN4XJA) so that it actually uses the Index and I've changed the search engine to `corax` as requested. I don't see any different results though.

**Reply by ayende** [reply]
Please note, your index won't actually work

```
Periods = schedule.Periods.Select(x => new
 {
     Day = x.Day,
     Start = x.Start,
     End = x.End,
     _ = CreateField("Periods_Day", x.Day)
 }).ToList()
````

The issue is that you are trying to index an array of complex objects.

You need to do this:


```
Periods_End = schedule.Periods.Select(x => x.End),
Periods_Start = schedule.Periods.Select(x => x.Start),
....
````

**Reply by dtaalbers** [reply]
@ayende  My apologies. I read over your comment before. I have updated the Index in [my example](https://dotnetfiddle.net/LN4XJA).

It already works better, but not totally I think. I think this is because of what you've said the problem is? When using this for the following `End` for the appointment (schedule is Wednesday from 08:30 to 12:00).

```
 End = new DateTime(2023, 10, 18, 12, 0, 0)
 ```
This will be found for both code and db side filtering. When changing the `End` to this:
```
 End = new DateTime(2023, 10, 18, 12, 0, 1)
 ```
Code side filtering doesn't find a schedule, which is OK. But the DB filtering side does find a schedule, which is not OK. Using the `Corax` search engine.

**Reply by garayx** [reply]
I guess its the same issue as discussed above, so it will be fixed as part of this ticket, https://issues.hibernatingrhinos.com/issue/RavenDB-21588

**Reply by dtaalbers** [reply]
Great, thank you!

**Comment by Lwiel** [comment]
Your code-side filtering takes each `Schedule` object, then iterates over its `Periods` list, and for each `SchedulePeriod` object checks `itemDay`, `itemStart` and `itemEnd`. This means you'll get a result only if one **particular** `SchedulePeriod` object passess all three checks at once.
![Screenshot 2023-11-21 162549](https://github.com/ravendb/ravendb/assets/32438625/8550c2fb-8059-455a-9d63-8e2b195097bd)

Your db-side filtering works in a different way. As a result of indexing, for each `Schedule` object you have `Days` list, `PeriodsEnd` list, and `PeriodsStart` list. You're checking if any element in each of these lists fulfills condition you provided.
![Screenshot 2023-11-21 163028](https://github.com/ravendb/ravendb/assets/32438625/05ce330c-cef7-4430-b2e5-e1c7225b54e1)

This means you're checking if any element in `PeriodsEnd` list is greater than or equal to `12:00:00:1`, and `17:00:00` fulfills this condition.

It's explained with more details [in our documentation](https://ravendb.net/docs/article-page/5.4/csharp/indexes/indexing-nested-data).
If you want to achieve same result as in code-side filtering, you should use [fanout indexes](https://ravendb.net/docs/article-page/6.0/csharp/indexes/fanout-indexes).

Example index could look like this
```csharp
public class Schedules_TimeOnly_Fanout : AbstractIndexCreationTask<Schedule, Schedules_TimeOnly_Fanout.Result>
    {
        public class Result
        {
            public string Id { get; set; }
            public string UserId { get; set; }
            public DayOfWeek Day { get; set; }
            public TimeOnly End { get; set; }
            public TimeOnly Start { get; set; }
        }
        public Schedules_TimeOnly_Fanout()
        {
            Map = schedules => from schedule in schedules
                from period in schedule.Periods
                select new
                {
                    schedule.Id,
                    schedule.UserId,
                    period.Day,
                    period.End,
                    period.Start
                };
            
            StoreAllFields(FieldStorage.Yes);
        }
    }
```
